File: ai-critic-react/src/api/customAgents.ts
export interface CustomAgent {
  id: string;
  name: string;
  description: string;
  mode: 'primary' | 'subagent';
  model?: string;
  tools: Record<string, boolean>;
  permissions?: Record<string, string>;
  hasSystemPrompt: boolean;
}

export interface CustomAgentDetail extends CustomAgent {
  systemPrompt?: string;
}

export interface CreateCustomAgentRequest {
  id?: string;
  name: string;
  description?: string;
  mode?: 'primary' | 'subagent';
  model?: string;
  tools?: Record<string, boolean>;
  permissions?: Record<string, string>;
  template?: string;
  systemPrompt?: string;
}

export interface UpdateCustomAgentRequest {
  name?: string;
  description?: string;
  mode?: 'primary' | 'subagent';
  model?: string;
  tools?: Record<string, boolean>;
  permissions?: Record<string, string>;
  systemPrompt?: string;
}

export interface LaunchCustomAgentRequest {
  projectDir: string;
}

export interface LaunchCustomAgentResponse {
  sessionId: string;
  port: number;
  url: string;
}

export interface AgentTemplate {
  id: string;
  name: string;
  description: string;
  mode: 'primary' | 'subagent';
  tools: Record<string, boolean>;
  permissions: Record<string, string>;
}

export const AGENT_TEMPLATES: AgentTemplate[] = [
  {
    id: 'build',
    name: 'Build',
    description: 'Full development agent with all tools enabled',
    mode: 'primary',
    tools: {
      write: true,
      edit: true,
      bash: true,
      grep: true,
      read: true,
      webfetch: true,
      websearch: true,
    },
    permissions: {},
  },
  {
    id: 'plan',
    name: 'Plan',
    description: 'Planning and analysis - read-only, no changes',
    mode: 'primary',
    tools: {
      read: true,
      grep: true,
      webfetch: true,
      websearch: true,
    },
    permissions: {
      edit: 'deny',
      bash: 'deny',
    },
  },
  {
    id: 'refactor',
    name: 'Refactor',
    description: 'Code refactoring specialist',
    mode: 'subagent',
    tools: {
      read: true,
      edit: true,
      write: true,
      grep: true,
    },
    permissions: {
      bash: 'deny',
    },
  },
  {
    id: 'debug',
    name: 'Debug',
    description: 'Debugging and investigation specialist',
    mode: 'subagent',
    tools: {
      read: true,
      grep: true,
      bash: true,
      webfetch: true,
    },
    permissions: {
      edit: 'ask',
      write: 'ask',
    },
  },
];

export async function fetchCustomAgents(): Promise<CustomAgent[]> {
  const resp = await fetch('/api/custom-agents');
  if (!resp.ok) {
    throw new Error('Failed to fetch custom agents');
  }
  return resp.json();
}

export async function fetchCustomAgent(id: string): Promise<CustomAgentDetail> {
  const resp = await fetch(`/api/custom-agents/${encodeURIComponent(id)}`);
  if (!resp.ok) {
    throw new Error('Failed to fetch custom agent');
  }
  return resp.json();
}

export async function createCustomAgent(
  agent: CreateCustomAgentRequest
): Promise<CustomAgent> {
  const resp = await fetch('/api/custom-agents', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(agent),
  });
  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error || 'Failed to create custom agent');
  }
  return resp.json();
}

export async function updateCustomAgent(
  id: string,
  agent: UpdateCustomAgentRequest
): Promise<CustomAgent> {
  const resp = await fetch(`/api/custom-agents/${encodeURIComponent(id)}`, {
    method: 'PUT',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(agent),
  });
  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error || 'Failed to update custom agent');
  }
  return resp.json();
}

export async function deleteCustomAgent(id: string): Promise<void> {
  const resp = await fetch(`/api/custom-agents/${encodeURIComponent(id)}`, {
    method: 'DELETE',
  });
  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error || 'Failed to delete custom agent');
  }
}

export async function launchCustomAgent(
  id: string,
  projectDir: string
): Promise<LaunchCustomAgentResponse> {
  const resp = await fetch(`/api/custom-agents/${encodeURIComponent(id)}/launch`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ projectDir }),
  });
  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error || 'Failed to launch custom agent');
  }
  return resp.json();
}

export interface CustomAgentSession {
  id: string;
  agent_id: string;
  agent_name: string;
  project_dir: string;
  port: number;
  created_at: string;
  status: string;
}

export async function fetchCustomAgentSessions(): Promise<CustomAgentSession[]> {
  const resp = await fetch('/api/custom-agents/sessions');
  if (!resp.ok) {
    const error = await resp.text();
    throw new Error(error || 'Failed to fetch custom agent sessions');
  }
  return resp.json();
}

File: client/custom_agents.go
package client

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"
)

type CustomAgent struct {
	ID              string            `json:"id"`
	Name            string            `json:"name"`
	Description     string            `json:"description"`
	Mode            string            `json:"mode"`
	Model           string            `json:"model,omitempty"`
	Tools           map[string]bool   `json:"tools"`
	Permissions     map[string]string `json:"permissions,omitempty"`
	HasSystemPrompt bool              `json:"hasSystemPrompt"`
	SystemPrompt    string            `json:"systemPrompt,omitempty"`
}

type CreateCustomAgentRequest struct {
	ID           string            `json:"id,omitempty"`
	Name         string            `json:"name,omitempty"`
	Description  string            `json:"description,omitempty"`
	Mode         string            `json:"mode,omitempty"`
	Model        string            `json:"model,omitempty"`
	Tools        map[string]bool   `json:"tools,omitempty"`
	Permissions  map[string]string `json:"permissions,omitempty"`
	Template     string            `json:"template,omitempty"`
	SystemPrompt string            `json:"systemPrompt,omitempty"`
}

type UpdateCustomAgentRequest struct {
	Name         string            `json:"name,omitempty"`
	Description  string            `json:"description,omitempty"`
	Mode         string            `json:"mode,omitempty"`
	Model        string            `json:"model,omitempty"`
	Tools        map[string]bool   `json:"tools,omitempty"`
	Permissions  map[string]string `json:"permissions,omitempty"`
	SystemPrompt string            `json:"systemPrompt,omitempty"`
}

type CustomAgentSession struct {
	ID         string `json:"id"`
	AgentID    string `json:"agent_id"`
	AgentName  string `json:"agent_name"`
	ProjectDir string `json:"project_dir"`
	Port       int    `json:"port"`
	CreatedAt  string `json:"created_at"`
	Status     string `json:"status"`
	Error      string `json:"error,omitempty"`
}

type LaunchCustomAgentResponse struct {
	SessionID string `json:"sessionId"`
	Port      int    `json:"port"`
	URL       string `json:"url"`
}

func (c *Client) ListCustomAgents() ([]CustomAgent, error) {
	var out []CustomAgent
	if err := c.getJSON("/api/custom-agents", &out); err != nil {
		return nil, err
	}
	if out == nil {
		out = []CustomAgent{}
	}
	return out, nil
}

func (c *Client) GetCustomAgent(agentID string) (*CustomAgent, error) {
	var out CustomAgent
	if err := c.getJSON("/api/custom-agents/"+url.PathEscape(agentID), &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *Client) CreateCustomAgent(reqBody CreateCustomAgentRequest) (*CustomAgent, error) {
	var out CustomAgent
	if err := c.postJSON("/api/custom-agents", reqBody, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *Client) UpdateCustomAgent(agentID string, reqBody UpdateCustomAgentRequest) (*CustomAgent, error) {
	var out CustomAgent
	if err := c.sendJSON(http.MethodPut, "/api/custom-agents/"+url.PathEscape(agentID), reqBody, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *Client) DeleteCustomAgent(agentID string) error {
	req, err := c.NewRequest(http.MethodDelete, "/api/custom-agents/"+url.PathEscape(agentID), nil)
	if err != nil {
		return err
	}
	resp, err := c.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return readAPIError(resp)
	}
	return nil
}

func (c *Client) ListCustomAgentSessions(agentID string) ([]CustomAgentSession, error) {
	var out []CustomAgentSession
	if err := c.getJSON("/api/custom-agents/"+url.PathEscape(agentID)+"/sessions", &out); err != nil {
		return nil, err
	}
	if out == nil {
		out = []CustomAgentSession{}
	}
	return out, nil
}

func (c *Client) StopCustomAgentSession(sessionID string) error {
	req, err := c.NewRequest(http.MethodDelete, "/api/custom-agents/sessions/"+url.PathEscape(sessionID), nil)
	if err != nil {
		return err
	}
	resp, err := c.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return readAPIError(resp)
	}
	return nil
}

func (c *Client) LaunchCustomAgent(agentID string, projectDir string, sessionID string) (*LaunchCustomAgentResponse, error) {
	reqBody := struct {
		ProjectDir string `json:"projectDir"`
		SessionID  string `json:"sessionId,omitempty"`
	}{
		ProjectDir: projectDir,
		SessionID:  sessionID,
	}

	var out LaunchCustomAgentResponse
	if err := c.postJSON("/api/custom-agents/"+url.PathEscape(agentID)+"/launch", reqBody, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (c *Client) WaitCustomAgentSessionReady(sessionID string, timeout time.Duration) error {
	deadline := time.Now().Add(timeout)
	path := "/api/custom-agents/sessions/" + url.PathEscape(sessionID) + "/proxy/session"

	for {
		req, err := c.NewRequest(http.MethodGet, path, nil)
		if err != nil {
			return err
		}
		resp, err := c.Do(req)
		if err == nil {
			resp.Body.Close()
			switch {
			case resp.StatusCode >= 200 && resp.StatusCode < 300:
				return nil
			case resp.StatusCode == http.StatusServiceUnavailable:
				// still starting
			case resp.StatusCode == http.StatusNotFound:
				return fmt.Errorf("session %s is not running", sessionID)
			default:
				return fmt.Errorf("wait for session ready: unexpected status %s", resp.Status)
			}
		}
		if time.Now().After(deadline) {
			return fmt.Errorf("timeout waiting for custom agent session %s to become ready", sessionID)
		}
		time.Sleep(500 * time.Millisecond)
	}
}

func (c *Client) postJSON(path string, reqBody any, out any) error {
	return c.sendJSON(http.MethodPost, path, reqBody, out)
}

func (c *Client) sendJSON(method string, path string, reqBody any, out any) error {
	data, err := json.Marshal(reqBody)
	if err != nil {
		return fmt.Errorf("marshal %s request: %w", path, err)
	}

	req, err := c.NewRequest(method, path, bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return readAPIError(resp)
	}
	if out == nil {
		return nil
	}
	if err := json.NewDecoder(resp.Body).Decode(out); err != nil {
		return fmt.Errorf("decode %s response: %w", path, err)
	}
	return nil
}
